src/fdc3/dotnet/DesktopAgent/src/DesktopAgent/Fdc3.cs

```diff
@@ -0,0 +1,51 @@
+﻿/*
+ * Morgan Stanley makes this available to you under the Apache License,
+ * Version 2.0 (the "License"). You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership. Unless required by applicable law or agreed
+ * to in writing, software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using MorganStanley.ComposeUI.Messaging;
+
+namespace MorganStanley.ComposeUI.Fdc3.DesktopAgent
+{
+    public class Fdc3 : IAsyncDisposable
+    {
+        private List<UserChannel> _userChannels = new();
+
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly IMessageRouter _messageRouter;
+
+        public Fdc3(IServiceProvider serviceProvider)
+        {
+            _messageRouter = serviceProvider.GetRequiredService<IMessageRouter>();
+            _loggerFactory = serviceProvider.GetService<ILoggerFactory>() ?? NullLoggerFactory.Instance;
+        }
+
+        public async ValueTask AddUserChannel(string id)
+        {
+            var uc = new UserChannel(id, _messageRouter, _loggerFactory.CreateLogger<UserChannel>());
+            await uc.Connect();
+            _userChannels.Add(uc);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            var tasks = _userChannels.Select(x => x.DisposeAsync()).ToArray();
+            foreach (var task in tasks)
+            {
+                await task;
```

Comment:
I think it would make sense to wrap these in a try/catch, so if disposing a channel throws an exception then that wouldn't prevent disposing the remaining.

Reply:
It wouldn't prevent disposing, the ToArray() is there to start them in parallel, but I added the try-catch to still wait for all of them to finish.